django/contrib/admindocs/utils.py

```diff
@@ -145,93 +146,53 @@ def default_reference_role(name, rawtext, text, lineno, inliner, options=None, c
     for name, urlbase in ROLES.items():
         create_reference_role(name, urlbase)
 
-# Match the beginning of a named or unnamed group.
-named_group_matcher = re.compile(r'\(\?P(<\w+>)')
-unnamed_group_matcher = re.compile(r'\(')
 
-
-def replace_named_groups(pattern):
-    r"""
-    Find named groups in `pattern` and replace them with the group name. E.g.,
-    1. ^(?P<a>\w+)/b/(\w+)$ ==> ^<a>/b/(\w+)$
-    2. ^(?P<a>\w+)/b/(?P<c>\w+)/$ ==> ^<a>/b/<c>/$
-    """
-    named_group_indices = [
-        (m.start(0), m.end(0), m.group(1))
-        for m in named_group_matcher.finditer(pattern)
-    ]
-    # Tuples of (named capture group pattern, group name).
-    group_pattern_and_name = []
-    # Loop over the groups and their start and end indices.
-    for start, end, group_name in named_group_indices:
-        # Handle nested parentheses, e.g. '^(?P<a>(x|y))/b'.
-        unmatched_open_brackets, prev_char = 1, None
-        for idx, val in enumerate(list(pattern[end:])):
-            # If brackets are balanced, the end of the string for the current
-            # named capture group pattern has been reached.
-            if unmatched_open_brackets == 0:
-                group_pattern_and_name.append((pattern[start:end + idx], group_name))
-                break
-
-            # Check for unescaped `(` and `)`. They mark the start and end of a
-            # nested group.
-            if val == '(' and prev_char != '\\':
-                unmatched_open_brackets += 1
-            elif val == ')' and prev_char != '\\':
-                unmatched_open_brackets -= 1
-            prev_char = val
-
-    # Replace the string for named capture groups with their group names.
-    for group_pattern, group_name in group_pattern_and_name:
-        pattern = pattern.replace(group_pattern, group_name)
-    return pattern
+# Match unescaped parenthesis in a regex, simplify groups and then clean.
+parenthesis_matcher = re.compile(r'\\.|([()])')
+groups_matcher = re.compile(r'^\((?:\?(?:([iLmsux:])|P(<\w+>)|([P#=!<(]))|(?#empty))(.*)\)$')
+cleanup_matcher = re.compile(
+    r'(\\0[0-9]{0,3})|\\([^1-9AbBdDsSwWZafnrt])|\\[1-9][0-9]?|\\[AbBdDsSwWZafnrt]|[.^$*+?|]|\[[^]]*\]|\{[^}]*\}')
```

Comment:
What is "1-9AbBdDsSwWZafnrt"? No tests are failing if some of those characters are removed, so probably more tests should be added.

Can you add some inline comments to the regex similar to:
https://github.com/django/django/blob/c930c241f8f1ccf3a7848b843628eacdb983d70a/django/utils/http.py#L22-L29